### ISSUE TITLE
Improved structure for Solar Time Varying Tariffs

### ISSUE THREAD

**Issue by DannyDuong** [ConsumerDataStandardsAustralia/standards-maintenance#624]
### Description
Time Varying Tariffs for Solar Feed-In Tariffs are relatively new and were introduced as part of EnergyPlanResponseV2.
In Victoria, one offer/plan, can have up to 4 solarFeedInTariff objects, to cater for different solar policies:
•	Premium FiT
•	Current FiT
•	Variable FiT – option 1
•	Variable FiT – option 2

It’s important for the data standards to facilitate all the items relating to a given Solar FiT policy in the same object.

For reference, in Victoria, the approved Feed-in Tariff structures and rates for 2023–24 are below:

![image](https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/assets/37787868/76ce3d08-3fcb-4b0b-b450-7c214e81816e)

Source: Essential Services Commission website, https://www.esc.vic.gov.au/electricity-and-gas/prices-tariffs-and-benchmarks/minimum-feed-tariff/minimum-feed-tariff-review-2023-24, accessed 16 May 2023.

### Area Affected
The EnergyPlanResponseV2 response payload for the Get Generic Plan Detail API.

The following are key requirements regarding standards for the Solar Tariff:

1. Each _solarFeedInTariff_ to be a full tariff, i.e. it covers 24hrs by 7 days for Solar FiT with multiple rates with different time periods.
2. In addition, all the items relating to this tariff should be kept in the same object – e.g. Overnight, Day and Early evening are different tariffs (_timeVaryingTariffs_) for the same variable Solar FiT object.

It is important, that the data standards can support an offer having multiple Solar FiT policies, and each Solar FiT policies may have multiple rates with different time periods. 
The current standard does not support this, and hence the need for this CR.

The work-around using the current standards may be possible, but it would lead to:
•	be a very confusing and incorrect way to represent _timeVaryingTariffs_, as they are not nested at the right level
•	could have a flow-on effect on anyone consuming the CDR offer data
•	result in duplication of implementation costs to develop with this work-around and then to correct it later

### Change Proposed
DEECA propose a CR to change the _timeVaryingTariffs_ from an object to be an array of objects to cater for the three set of timebands and rates.
This change will facilitate the data standards to support an offer having multiple Solar FiT policies, and each Solar FiT policies may have multiple rates with different time periods. 

In addition, a _timeVaryingTariffs.displayName_ field is required to describe each object in the _timeVaryingTariffs_ array. 
Note: This is separate from the _timeVaryingTariffs.type_ field which is a generic Peak, Off-Peak and Shoulder

Refer to the following sample payload
[solarFeedInTariff_example_nested.jsonc](https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/files/13608527/solarFeedInTariff_example_nested.jsonc)

# DSB Proposed Solution

The proposed solution can be found through the staging link provided in https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/624#issuecomment-2019888328

**Comment by mattyp** [comment]
+1

I see [an original proposal for the `timeVaryingTariffs`](https://github.com/ConsumerDataStandardsAustralia/standards/issues/190#issuecomment-927507635) has it as an array, so I am guessing it was accidently converted to a single object somewhere along the implementation journey.

Also doesn't make sense to have a plural name for only a single value 😛

**Comment by mattyp** [comment]
PS - duplicate of https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/456#issue-1097434652

**Comment by mattyp** [comment]
Option 1 ✅. It’s a shame a bug fix is a breaking change, but best to just fix it up, get it right and move onto the next bug.

**Comment by jsn2510** [comment]
The AER support option 1.

**Comment by DannyDuong** [comment]
DEECA proposed and hence support option 1

**Comment by DannyDuong** [comment]
@HemangCDR I've included a revised file and have pasted the content below - which should replace the Option 1 that you have posted above.

It is largely the same as the initial file that I've shared, but for completeness, I've included the "period" and "volume" fields to ensure that it supports the step tariff linked to this ticket here https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/625

"solarFeedInTariff" : [
    {
        "displayName": "Time-Varying - Option 1",
        "description": "Victorian time-varying FIT",
        "scheme": "Variable FiT",
        "payerType": "GOVERNMENT",
        "tariffUType": "timeVaryingTariffs",
        "period": "string",
        "timeVaryingTariffs": [
		{
            "type": "PEAK", 
			"displayName": "Overnight",
            "rates": [
                {
                    "unitPrice": "0.113",
                    "measureUnit": "KWH",
                    "volume": 0
                }
            ],
            "timeVariations": [
                {
                    "days": ["MON","TUE","WED","THU","FRI","SAT","SUN"],
                    "startTime": "22:00", 
                    "endTime": "07:00" 
                }
            ]
        },
		{
            "type": "OFFPEAK", 
			"displayName": "Day",
            "rates": [
                {
                    "unitPrice": "0.044",
                    "measureUnit": "KWH",
                    "volume": 0
                }
            ],
            "timeVariations": [
                {
                    "days": ["MON","TUE","WED","THU","FRI"],
                    "startTime": "07:00", 
                    "endTime": "15:00"
                },
                {
                    "days": ["MON","TUE","WED","THU","FRI"],
                    "startTime": "21:00",
                    "endTime": "22:00" 
                },
                {
                    "days": [
                        "SAT",
                        "SUN"
                    ],
                    "startTime": "07:00", 
                    "endTime": "22:00" 
                }
            ]
        },
		{
            "type": "SHOULDER", 
	    "displayName": "Early Evenning",
            "rates": [
                {
                    "unitPrice": "0.093",
                    "measureUnit": "KWH",
                    "volume": 0
                }
            ],
            "timeVariations": [
                {
                    "days": ["MON","TUE","WED","THU","FRI"],
                    "startTime": "15:00", //assuming HH:MM format
                    "endTime": "21:00" //assuming HH:MM format
                }
            ]
        }
		]
    }
]


[solarFeedInTariff_example - DEECA Feb 2024.jsonc](https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/files/14414649/solarFeedInTariff_example.-.DEECA.Feb.2024.jsonc)

**Comment by AGL-CDR** [comment]
Should this item be accepted, we will need at least six months lead time, starting from the date that final standards are published. With that in mind, we propose a FDO of early December.

**Comment by HemangCDR** [comment]
Noting the feedback recieved thus far, and as per the action noted during the MI call on March 20, below is the proposed solution (i.e. option 1 in this comment) for this CR:

It consists of the following changes:

1. Convert `solarFeedInTariff.timeVaringTariffs` into an array to allow representation of mulitple feed in tariff for a given plan
2. Incorporate the new tariff types noted in the essential services review by extending the `solarFeedInTariff.timeVaryingTariffs.type` ENUM. 
3. The FDO date of 11/11/2024 is recommended, which is the last obligation date noted in the [obligation date schedule](https://consumerdatastandardsaustralia.github.io/standards/includes/endpoint-version-schedule/#obligation-dates-schedule). This is in alignment with the FDO of other energy CRs being discussed in this MI.

The changes would appear as below:
```json
"solarFeedInTariff": [
  {
    "displayName": "string",
    "description": "string",
    "startDate": "string",
    "endDate": "string",
    "scheme": "PREMIUM",
    "payerType": "GOVERNMENT",
    "tariffUType": "singleTariff",
    "singleTariff": {
      "rates": [
        {
          "unitPrice": "string",
          "measureUnit": "KWH",
          "volume": 0
        }
      ]
    },
    "timeVaryingTariffs": [ //converted into array of objects
        {
          "type": ["PEAK", "OFF_PEAK", "SHOULDER", "OVERNIGHT", "DAY", "EARLY_EVENING"], //New ENUM values for new tariff types noted in the essential services review
          "rates": [
            {
              "unitPrice": "string",
              "measureUnit": "KWH",
              "volume": 0
            }
          ],
          "timeVariations": [
            {
              "days": [
                "SUN"
              ],
              "startTime": "string",
              "endTime": "string"
            }
          ]
        }
    ]
  }
]
```

Feedback on the above is welcome.

**Comment by biza-io** [comment]
Biza.io supports the proposal most recently posted by @HemangCDR as we believe it is in the best interests of the ecosystem.

**Comment by CDR-API-Stream** [comment]
The change has been staged and can be reviewed here: https://github.com/ConsumerDataStandardsAustralia/standards-staging/compare/release/1.30.0...maintenance/624

**Note:** The retirement date for the following endpoints impacted as a result of this change has been set to **September 8th 2025**
- Get Generic Plan Detail v2
- Get Energy Account Detail v3

**Comment by DannyDuong** [comment]
> Noting the feedback recieved thus far, and as per the action noted during the MI call on March 20, below is the proposed solution (i.e. option 1 in this comment) for this CR:
> 
> It consists of the following changes:
> 
> 1. Convert `solarFeedInTariff.timeVaringTariffs` into an array to allow representation of multiple feed in tariff for a given plan
>    ~2. Incorporate the new tariff types noted in the essential services review by extending the `solarFeedInTariff.timeVaryingTariffs.type` ENUM.~
> 2. Add new `displayName` field as per original proposal. This is based on feedback received during MI call on 3rd April noting that it aligns with other time of use structures and gives DHs the flexibility as the use custom names.
> 3. The FDO date of 11/11/2024 is recommended, which is the last obligation date noted in the [obligation date schedule](https://consumerdatastandardsaustralia.github.io/standards/includes/endpoint-version-schedule/#obligation-dates-schedule). This is in alignment with the FDO of other energy CRs being discussed in this MI.
> 
> The changes would appear as below:
> 
> ```json
> "solarFeedInTariff": [
>   {
>     "displayName": "string",
>     "description": "string",
>     "startDate": "string",
>     "endDate": "string",
>     "scheme": "PREMIUM",
>     "payerType": "GOVERNMENT",
>     "tariffUType": "singleTariff",
>     "singleTariff": {
>       "rates": [
>         {
>           "unitPrice": "string",
>           "measureUnit": "KWH",
>           "volume": 0
>         }
>       ]
>     },
>     "timeVaryingTariffs": [ //converted into array of objects
>         {
>           "type": ["PEAK", "OFF_PEAK", "SHOULDER"],
>           "displayName" : "string", // new mandatory field
>           "rates": [
>             {
>               "unitPrice": "string",
>               "measureUnit": "KWH",
>               "volume": 0
>             }
>           ],
>           "timeVariations": [
>             {
>               "days": [
>                 "SUN"
>               ],
>               "startTime": "string",
>               "endTime": "string"
>             }
>           ]
>         }
>     ]
>   }
> ]
> ```
> 
> Feedback on the above is welcome.

@HemangCDR The following are DEECA feedback on this.

1. "scheme": ["PREMIUM", "OTHER", "CURRENT", "VARIABLE"],   // Adding CURRENT and VARIABLE for clarity, otherwise they will be treated as OTHER
2. "period": "string",   // the period field should be added here for completeness, at the same level as tariffUType. This is link to https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/625 
3. Unsure what purpose the startDate and endDate fields; at the moment, DEECA does not know how to populate them. These fields should be removed, as they are not in the scope of this CR. If anything, they are related to the "period" field for step tariff calculations rather than the Solar Time Varying Tariffs per se and therefore should be discussed separately
4. I'm assuming that the standards nesting structure support the example "Time-Varying - Option 1" that was provided above

**Comment by CDR-API-Stream** [comment]
@DannyDuong thank you for your comment. Please note the following response:

> 1. "scheme": ["PREMIUM", "OTHER", "CURRENT", "VARIABLE"],   // Adding CURRENT and VARIABLE for clarity, otherwise they will be treated as OTHER

This is a good pickup and can be included as part of this change noting that this was implied in the original proposal. 

> 2. "period": "string",   // the period field should be added here for completeness, at the same level as tariffUType. This is link to [Additional field to support Step Tariff calculations #625](https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/625)

This has been already been addressed as part of #625

> 3. I'm assuming that the standards nesting structure support the example "Time-Varying - Option 1" that was provided above

That is correct.

The above was discussed in the MI19 call today (17th April) and participants supported introducing the two ENUM values as part of this change and MI18 for Chair approval. The staged changes will be updated to reflect this.

**Comment by nils-work** [comment]
Standards version 1.30.0 was published on 24/04/2024 incorporating this change from MI18.
